src/engine/renderer/tr_shader.cpp

```diff
@@ -5120,30 +5145,54 @@ static void CollapseStages()
 		}
 	}
 
-	// move all active stages at beginning
-	// note that the 'active' field is still used instead of numStages in some code that runs later
-	int numActiveStages = 0;
+	bool lightStageFound = false;
+
 	for ( int i = 0; i < MAX_SHADER_STAGES; i++ )
 	{
-		if ( !stages[ i ].active )
+		shaderStage_t *stage = &stages[ i ];
+
+		if ( !stage->active )
 		{
 			continue;
 		}
 
-		if ( i != numActiveStages )
+		if ( r_showLightMaps->integer && lightStageFound )
 		{
-			stages[ numActiveStages ] = stages[ i ];
-
-			stages[ i ].active = false;
+			stage->active = false;
+			continue;
 		}
 
-		numActiveStages++;
+		switch ( stage->type )
+		{
+			case stageType_t::ST_LIQUIDMAP:
+				if ( !r_liquidMapping->integer )
+				{
+					stage->type = stageType_t::ST_COLLAPSE_DIFFUSEMAP;
+					stage->bundle[ TB_DIFFUSEMAP ].image[ 0 ] = tr.whiteImage;
+				}
+				break;
+			case stageType_t::ST_REFLECTIONMAP:
+			case stageType_t::ST_COLLAPSE_REFLECTIONMAP:
+				if ( !r_reflectionMapping->integer )
+				{
+					stage->active = false;
+				}
+				break;
+			case stageType_t::ST_LIGHTMAP:
+			case stageType_t::ST_DIFFUSEMAP:
+			case stageType_t::ST_COLLAPSE_DIFFUSEMAP:
+				lightStageFound = true;
+				break;
+			default:
+				break;
+		}
 	}
 
-	shader.numStages = numActiveStages;
-
-	// Do some precomputation.
+	PileUpStages();
```

Comment:
Add comment here describing what we're doing.

I know you kinda commented this in the function, but let's also comment the flow of the shader parsing code because it can be quite complex.

Reply:
I'm also looking for a better function name.

Basically it turns this:

```c
maxStages = 10;
stages[maxStages] = { active, active, inactive, active, inactive, inactive, active, null, null, null };
numStages = 7;
```

into this:

```c
maxStages = 10;
stages[maxStages] = { active, active, active, active, inactive, inactive, inactive, null, null, null };
numStages = 4;
```

That meaning loops iterating stages don't have to do this:

```c
for ( i = 0; i < 7; i++ )
{
  if ( stages[ i ] == inactive )
    continue;

  render( stage[ i ] );
}
```

but can do this:

```c
for ( i = 0; i < 5; i++ )
{
  render( stage[ i ] );
}
```

For reference our current code in master is doing this overly complicated code:

```c
for ( i = 0; i < 10; i++ )
{
  if ( stages[ i ] == inactive )
    continue;

  if ( stages[ i ] == null )
    break;

  render( stage[ i ] );
}
```

Reply:
So I look for a function name that makes obvious we are compacting the stage list.

Reply:
SortStages(), GroupActiveStages() is what comes to mind for a function name.